streams/src/main/java/org/apache/kafka/streams/kstream/Windows.java

```diff
@@ -39,17 +43,48 @@
     private long maintainDurationMs = 24 * 60 * 60 * 1000L; // default: one day
     @Deprecated public int segments = 3;
 
+    private Duration grace;
+
     protected Windows() {}
 
+    /**
+     * Reject late events that arrive more than {@code afterWindowEnd}
+     * after the end of its window.
+     *
+     * @param afterWindowEnd The grace period to admit late-arriving events to a window.
+     * @return this updated builder
+     */
+    public Windows<W> grace(final Duration afterWindowEnd) {
+        if (afterWindowEnd.isNegative()) {
+            throw new IllegalArgumentException("Grace period must not be negative.");
+        }
+
+        grace = validateMillisecondDuration(afterWindowEnd, "Grace period must be expressible in milliseconds");
+
+        return this;
+    }
+
+    /**
+     * Return the window grace period (the time to admit
+     * late-arriving events after the end of the window.
+     */
+    @SuppressWarnings("deprecation") // continuing to support Windows#maintainMs/segmentInterval in fallback mode
+    public Duration grace() {
+        return grace != null ? grace : Duration.ofMillis(maintainMs() - size());
+    }
+
     /**
      * Set the window maintain duration (retention time) in milliseconds.
      * This retention time is a guaranteed <i>lower bound</i> for how long a window will be maintained.
      *
      * @param durationMs the window retention time in milliseconds
      * @return itself
      * @throws IllegalArgumentException if {@code durationMs} is negative
+     * @deprecated since 2.1. Use {@link Joined#withRetention(Duration)} or {@link Materialized#withRetention(Duration)}
```

Comment:
Hmm, I'm not sure where to put such a comment, since Windows is now ignorant of retention period and Materialized/StateStore are ignorant of grace period.

I think I should instead place this remark in the release notes. WDYT? In fact, should I go ahead an include the release notes docs in this PR?

Reply:
I think leaving this in Windows is fine, in which we refer to `Materialized#retention`.